### DISCUSSION TITLE
windows下使用了nvm，用户目录下配置~/.czrc，自定义配置不生效

### DISCUSSION THREAD

**Discussion by WakerCN** [Zhengqbbb/cz-git#155] [Q&A]
# 背景
想要全局使用`cz`，但发现无法正常加载`~/.czrc`配置文件

# 环境
* window11, node18.17.0
* 使用了nvm
* nvm Node目录进行了更改,更改到了其他磁盘路径，不在c盘，nvm路径配置如下图
![image](https://github.com/Zhengqbbb/cz-git/assets/50485199/94f84697-dac4-46b9-af3f-98e654d52015)

此时编辑~/.czrc，自定义配置不生效
`.czrc`内容如下代码，不知道是不是和path字段值有关
``` json
{
  "path": "cz-git",
  "$schema": "https://cdn.jsdelivr.net/gh/Zhengqbbb/cz-git@1.8.0/docs/public/schema/cz-git.json",
  ...
}
```

# 其他问题：
* 全局安装后，无法找到`czg`命令
看到文档中说明
![image](https://github.com/Zhengqbbb/cz-git/assets/50485199/731d2f70-9754-41ff-99f2-c0c1ac4fa122)
但是在用户目录下并没有找到 `.zshrc`和`.bashrc`文件
![image](https://github.com/Zhengqbbb/cz-git/assets/50485199/002db1a9-e7f4-472f-a56e-c4ba4267364d)

这种情况，应该怎么处理

**Comment by Zhengqbbb** [comment]
## 配置未生效：
这里主要是 windows 环境中 Node 的 `$HOME` 目录即 `~` 根目录的**到底在哪**。
> 你的潜意识告诉你，在 `C` 盘

但从图中以及之前的经验可以看出可能是 D盘的根目录 `D://`，参考这个 https://github.com/Zhengqbbb/cz-git/discussions/132

这里有几个命令可以提供给你 DEBUG，你可以尝试一下结合自己的环境，毕竟我手中也没有 Windows 可以复现：

```sh
node -e '
const os = require("node:os")
console.log(os.homedir())
'
```

```sh
# 设置环境变量
CZ_DEBUG=1 cz
# Windows powershell 似乎是这样
cz -CZ_DEBUG 1
```

---

## 无法找到命令
是的，nvm ，环境版本管理器，全局安装包目录未在系统变量 `path` 中设置 ，**出于对各个版本的完美隔离**，并不会帮你手动设置从 npm 全局下载的包路径即 `$PATH`
不能说只有 `czg` 命令，应该是你全部的全局下载的包你现在都无法直接启动
需要用户手动决定，这个可以上网搜一下 "npm Windows 设置系统环境变量 path "。

> 其实一个命令之所以可以在终端通过名字的方式直接调用，关键就是在这些包的目录有没有在全局的 `$PATH` 之中
> 类似这个：https://juejin.cn/post/6946594249093677069#heading-6 。在 Windows 中设置 `$PATH` 没办法通过命令的方式，或者说没右键设置这么简单。我记得以前在 Windows 中设置 `JAVA_HOME` 总是很烦

- [x] nvm 会帮你设置总 node 安装路径
- [x] nvm 会帮你设置各个版本的 node 安装路径
- [ ] nvm 不会帮你设置各个版本的 node 安装路径下的 npm 全局包路径

**全局下载可执行文件路径如何查找**：`npm -g prefix` 的 bin 路径，需要把这个路径设置到 `$PATH` 中，这样命令行直接输入才能找到可执行文件 `czg`。

<img width="1090" alt="image" src="https://github.com/Zhengqbbb/cz-git/assets/40693636/617b2eb0-bc47-4726-8fd2-24eea8ee0759">


---

你使用的 Windows PowerShell 所以 `.zshrc` `.bashrc` 不会生效的

**Comment by WakerCN** [comment]
![image](https://github.com/Zhengqbbb/cz-git/assets/50485199/5798eaeb-398f-4b7a-85d6-1939559853ff)
这里可以看到其实环境变量里面path是有node的bin目录的

其他全局安装的包是可以在全局使用命令，这一点可以验证我的环境变量应该是设置没问题

有个疑问，在全局node目录下其他命令都会自动生成对应bin文件，但是全局安装了cz-git却没有生成对应的bin文件
![image](https://github.com/Zhengqbbb/cz-git/assets/50485199/be27a792-687b-4214-97d7-cdd3a23d8c16)

但在全局确实已经有了cz-git的包
![image](https://github.com/Zhengqbbb/cz-git/assets/50485199/9b742800-28a8-4f4d-828c-ac29381e449e)

然后我查看了cz-git包中的`package.json`文件，发现里面没有`bin`字段
我了解到的知识是安装包的如果package.json中有bin字段才会自动生成对应的bin文件，不知道是不是这个原因

**Reply by Zhengqbbb** [reply]
!!!!

```sh
npm install -g commitizen
```

or

```sh
npm install -g czg
```

---

cz-git 和 czg 是两个包：!!!
- `cz-git`: 是 [commitizen](https://github.com/commitizen/cz-cli) cz-cli 的交互插件，你配置当中的 path 就是指，当你使用 `cz`或 `git cz` 启动时，你要用什么插件
- `czg`: 你可以理解为完全重写的 commitizen + cz-git，独立 0依赖的命令行工具

https://cz-git.qbb.sh/zh/cli/why#%E5%8A%A8%E6%9C%BA

![CleanShot 2024-01-31 at 10 46 48@2x](https://github.com/Zhengqbbb/cz-git/assets/40693636/49e5ee56-9060-49f2-88dd-a0f4858f6e21)
